index.js

```diff
@@ -1,10 +1,97 @@
 'use strict'
 
 const From = require('fastify-reply-from')
-const WebSocketPlugin = require('fastify-websocket')
 const WebSocket = require('ws')
-const { pipeline } = require('stream')
-const nonWsMethods = ['DELETE', 'HEAD', 'PATCH', 'POST', 'PUT', 'OPTIONS']
+
+const httpMethods = ['DELETE', 'GET', 'HEAD', 'PATCH', 'POST', 'PUT', 'OPTIONS']
+
+function liftErrorCode (code) {
+  if (typeof code !== 'number') {
+    // Sometimes "close" event emits with a non-numeric value
+    return 1011
+  } else if (code === 1004 || code === 1005 || code === 1006) {
+    // ws module forbid those error codes usage, lift to "application level" (4xxx)
+    return 4000 + (code % 1000)
+  } else {
+    return code
+  }
+}
+
+function closeWebSocket (socket, code, reason) {
+  if (socket.readyState === WebSocket.OPEN) {
+    socket.close(liftErrorCode(code), reason)
+  }
+}
+
+function waitConnection (socket, write) {
+  if (socket.readyState === WebSocket.CONNECTING) {
+    socket.once('open', write)
+  } else {
+    write()
+  }
+}
+
+function proxyWebSockets (source, target) {
+  function close (code, reason) {
+    closeWebSocket(source, code, reason)
+    closeWebSocket(target, code, reason)
+  }
+
+  source.on('message', data => waitConnection(target, () => target.send(data)))
+  source.on('ping', data => waitConnection(target, () => target.ping(data)))
+  source.on('pong', data => waitConnection(target, () => target.pong(data)))
+  source.on('close', close)
+  source.on('error', error => close(1011, error.message))
+  source.on('unexpected-response', () => close(1011, 'unexpected response'))
+
+  // source WebSocket is already connected because it is created by ws server
+  target.on('message', data => source.send(data))
+  target.on('ping', data => source.ping(data))
+  target.on('pong', data => source.pong(data))
+  target.on('close', close)
+  target.on('error', error => close(1011, error.message))
+  target.on('unexpected-response', () => close(1011, 'unexpected response'))
+}
+
+function createWebSocketUrl (options, request) {
+  const source = new URL(request.url, 'http://127.0.0.1')
+
+  const target = new URL(
+    options.rewritePrefix || options.prefix || source.pathname,
+    options.upstream
+  )
+
+  target.search = source.search
+
+  return target
+}
+
+function setupWebSocketProxy (fastify, options) {
+  const server = new WebSocket.Server({
+    path: options.prefix,
+    server: fastify.server,
+    ...options.wsServerOptions
+  })
+
+  fastify.addHook('onClose', (instance, done) => {
+    for (const client of server.clients) {
+      client.close()
+    }
+    done()
+  })
```

Comment:
now that this is responsibility of this plugin, we need a test for this logic.

Reply:
Added close event await to WebSocket tests.